code/5911-goce-lazarevski/script.js

```diff
@@ -1,3 +1,156 @@
+                    ${episode.cast.map(castMember => `${castMember.actor} (${castMember.character})`).join(', ')}
+                </div>
+                <div class="episode-cell plot-preview">${episode.plot}</div>
+            </div>
+        `;
+        episodesList.innerHTML += episodeRow;
+    });
+};
+
+const applyFilters = (data) => {
+    const nameFilter = document.querySelector("#name-filter").value.toLowerCase();
+    const eraFilter = document.querySelector("#era-filter").value;
+    const doctorFilter = document.querySelector("#doctor-filter").value;
+    const companionFilter = document.querySelector("#companion-filter").value;
+    
+    const filteredEpisodes = data.filter(episode => {
+        const matchesName = episode.title.toLowerCase().includes(nameFilter);
+        const matchesEra = !eraFilter || episode.era === eraFilter;
+        const matchesDoctor = !doctorFilter || episode.doctor.actor === doctorFilter;
+        const matchesCompanion = !companionFilter || (episode.companion && episode.companion.actor === companionFilter);
+        return matchesName && matchesEra && matchesDoctor && matchesCompanion;
+    });
+    
+    displayListOfEpisodes(filteredEpisodes);
+};
+
+const sortEpisodes = (data, key, order) => {
+    return data.slice().sort((a, b) => {
+        let valueA, valueB;
+        if (key === 'era') {
+            const eraOrder = { "Classic": 1, "Modern": 2, "Recent": 3 };
+            valueA = eraOrder[a.era];
+            valueB = eraOrder[b.era];
+        } else if (key === "cast") {
+            valueA = a.cast.length;
+            valueB = b.cast.length;
+        } else if (key === "broadcast") {
+            valueA = new Date(a.broadcast_date).getFullYear();
+            valueB = new Date(b.broadcast_date).getFullYear();
+        } else if (key === "doctor") {
+            valueA = a.doctor.actor ? a.doctor.actor.toLowerCase() : '';
+            valueB = b.doctor.actor ? b.doctor.actor.toLowerCase() : '';
+        } else {
+            valueA = a[key] ? a[key].toString().toLowerCase() : '';
+            valueB = b[key] ? b[key].toString().toLowerCase() : '';
+        }
+        return (valueA > valueB ? 1 : valueA < valueB ? -1 : 0) * (order === 'asc' ? 1 : -1);
+    });
+};
+
+
+
+document.addEventListener("DOMContentLoaded", async () => {
+    const data = await fetchData();
+    displayListOfEpisodes(data);
+    populateFilters(data);
+    
+    document.querySelector("#name-filter").addEventListener("input", () => applyFilters(data));
+    document.querySelector("#era-filter").addEventListener("change", () => applyFilters(data));
+    document.querySelector("#doctor-filter").addEventListener("change", () => applyFilters(data));
+    document.querySelector("#companion-filter").addEventListener("change", () => applyFilters(data));
+    
+    document.querySelectorAll(".header-cell").forEach(header => {
+        header.dataset.order = "asc";
```

Comment:
This code only executes once 😄 Otherwise it is great

Reply:
Yeah, it need to be refreshed to work for the other functionalities, I will try to fix it and commit the new version.